# Analysis of Car Sales on German eBay website
In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. We'll work here with the version modified by Dataquest. Their modifications include:

- Sampling of 50 000 data points from the full dataset
- Dirtying the dataset a bit to more closely resemble what we would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.


The aim of this project is to clean the data and get some ideas of the criteria that help define the cars price.

## Summary Table
**[1 Overview and Columns Cleaning](#intro) <br>
  [2 Analysis of the Price and the Mileage Columns](#section2) <br>
  [3 Selling Dates and Age of the Cars](#section3) <br>
  [4 Exploration of the Car Brands](#section4) <br>
  [5 Impact of the Damaged Status on the Price](#section5) <br>
  [6 Conclusion](#conclusion) <br>**

<a id='intro'></a>
# 1 Overview and Columns Cleaning

In [90]:
import pandas as pd
import numpy as np

In [91]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [92]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive. Mainly, we'll make the following edits to columns names:
- `yearOfRegistration` to `registration_year`
- `monthOfRegistration` to `registration_month`
- `notRepairedDamage` to `unrepaired_damage`
- `dateCreated` to `ad_created`

In [93]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [94]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [95]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Few observations:
- Almost all rows have the value "privat" in the `seller` column and "Angebot" in the `offer_type` column. We can remove both columns.
- The `registration_year` column has some inconsistent values. The min value is 1000, before the first car existed. The max value is 9999, a date in the future. So this column will need some cleaning.
- The price and odometer columns are not identified as numeric values. We'll convert them just below.
- The `num_photos` has a single value: 0. We'll remove it.

In [96]:
autos.drop(columns = ['seller', 'offer_type', 'num_photos'])

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       $5,000  control          bus               2004    manuell       158   
1       $8,500  control    limousine               1997  automatik       286   
2       $8,990     test    limousine               2009    manuell       102   
3       $4,350  control   kleinwagen               2007  automatik        71   
4       $1,350     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995  $24,900  control    limousine               2011  automatik       239   
49996   $1,980  control       cabrio               1996    manuell        75   
49997  $13,200     test       cabrio               2014  automatik        69   
49998  $22,900  control        kombi               2013    manuell       150   
49999   $1,250  control    limousine               1996    manuell       101   

        model   odometer  registration_month fuel_type       brand  \
0      andere  150,000km                   3       lpg     peugeot   
1         7er  150,000km                   6    benzin         bmw   
2        golf   70,000km                   7    benzin  volkswagen   
3      fortwo   70,000km                   6    benzin       smart   
4       focus  150,000km                   7    benzin        ford   
...       ...        ...                 ...       ...         ...   
49995      q5  100,000km                   1    diesel        audi   
49996   astra  150,000km                   5    benzin        opel   
49997     500    5,000km                  11    benzin        fiat   
49998      a3   40,000km                  11    diesel        audi   
49999  vectra  150,000km                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

<a id='section2'></a>
# 2 Analysis of the Price and the Mileage Columns

We'll convert both columns to numeric type.

In [97]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(float)
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(float)
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [98]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [99]:
# exploring price and odometer_km columns
autos[['price','odometer_km']].describe()

price    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000

Few observations:
- The `price` column has a minimum value of \\$0, which is possible in auctions. However, the max value of \\$100 000 000 is very odd. We suspect some outliers in this column, but not so many, as the third quartile has a reasonable value of \\$7 200

- The `odometer_km` column has the median, third quartile and max value equal, which needs explanation. Maybe mileages above 150 000 km appear as with this max value. <br> <br>

Let's have a closer look at the `price` column now.

In [100]:
# how many unique values has the price column?
autos['price'].unique().shape

(2357,)

In [101]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

Let's remove the cars that have a price above \\$300 000.

In [102]:
autos = autos[autos["price"].between(1,30000)]
autos["price"].describe()

count    47785.000000
mean      5175.226745
std       5695.476950
min          1.000000
25%       1200.000000
50%       2999.000000
75%       7000.000000
max      30000.000000
Name: price, dtype: float64

Now, let's try to understand the quartiles issue in the `odometer_km` column.

In [103]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].unique())

(13,)
[150000.  70000.  50000.  80000.  10000. 125000.  90000.  20000.  60000.
   5000.  30000. 100000.  40000.]


Only 13 values for 50 000 cars! Now we understand that the mileage is given by ranges.

In [104]:
autos['odometer_km'].value_counts()

150000.0    31343
125000.0     5012
100000.0     2067
90000.0      1692
80000.0      1372
70000.0      1165
60000.0      1096
50000.0       936
5000.0        791
40000.0       744
30000.0       702
20000.0       659
10000.0       206
Name: odometer_km, dtype: int64

So, basically, most of the cars have a mileage above 150 000 km. That's what it means.

<a id='section3'></a>
# 3 Selling Dates and Age of the Cars

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler

- `last_seen`: added by the crawler

- `ad_created`: from the website

- `registration_month`: from the website

- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:


In [105]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We notice that the first 10 characters represent the day (e.g. 2016-03-12). <br>
We'll calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen` columns (all string columns) as percentages.

In [106]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025364
2016-03-06    0.014105
2016-03-07    0.036078
2016-03-08    0.033379
2016-03-09    0.033107
2016-03-10    0.032207
2016-03-11    0.032542
2016-03-12    0.037166
2016-03-13    0.015674
2016-03-14    0.036622
2016-03-15    0.034258
2016-03-16    0.029716
2016-03-17    0.031705
2016-03-18    0.012891
2016-03-19    0.034697
2016-03-20    0.037878
2016-03-21    0.037229
2016-03-22    0.032814
2016-03-23    0.032228
2016-03-24    0.029256
2016-03-25    0.031516
2016-03-26    0.032395
2016-03-27    0.030951
2016-03-28    0.034760
2016-03-29    0.034237
2016-03-30    0.033714
2016-03-31    0.031788
2016-04-01    0.033567
2016-04-02    0.035325
2016-04-03    0.038694
2016-04-04    0.036622
2016-04-05    0.012975
2016-04-06    0.003139
2016-04-07    0.001402
Name: date_crawled, dtype: float64

We are dealing here with car sales in the month of March and beginning of April 2016. <br>There seems to be an almost uniform distirbution for the `date_crawled` column, except for the last two days which have a much lower frequency.

In [107]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
                ...   
2016-04-03    0.038924
2016-04-04    0.037020
2016-04-05    0.011677
2016-04-06    0.003202
2016-04-07    0.001277
Name: ad_created, Length: 75, dtype: float64

In [108]:
autos['ad_created'].str[:10].value_counts(dropna=False).sort_index()

2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
              ... 
2016-04-03    1860
2016-04-04    1769
2016-04-05     558
2016-04-06     153
2016-04-07      61
Name: ad_created, Length: 75, dtype: int64

The first ad created dates from August 2015. There used to be roughly one ad creation per month. <br>
The auction website started to be quite active in March 2016.

In [109]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001088
2016-03-06    0.004374
2016-03-07    0.005420
2016-03-08    0.007534
2016-03-09    0.009689
2016-03-10    0.010798
2016-03-11    0.012514
2016-03-12    0.024024
2016-03-13    0.008999
2016-03-14    0.012682
2016-03-15    0.015988
2016-03-16    0.016449
2016-03-17    0.028314
2016-03-18    0.007345
2016-03-19    0.015967
2016-03-20    0.020718
2016-03-21    0.020718
2016-03-22    0.021471
2016-03-23    0.018646
2016-03-24    0.019860
2016-03-25    0.019358
2016-03-26    0.016825
2016-03-27    0.015758
2016-03-28    0.021053
2016-03-29    0.022559
2016-03-30    0.024987
2016-03-31    0.023899
2016-04-01    0.022957
2016-04-02    0.025008
2016-04-03    0.025364
2016-04-04    0.024736
2016-04-05    0.123742
2016-04-06    0.220592
2016-04-07    0.130564
Name: last_seen, dtype: float64

The distribution looks uniform in March-April like the crawling date column, except for the last three days. It may be due to the end of the crawling period but it wouldn't indicate the car sale date. <br> <br>
Now, let's deal with the `registration_year` column, where we alreday mentioned some odd values (min of 1000 and max of 9999).
Because it would be unlikely to sell cars which are more than 20 years old, it would be a good reasonable assumption that the oldest car sold was registered in 1996. So let's remove cars that fall outside the 1996 - 2016 interval.

In [110]:
autos['registration_year'].describe()

count    47785.000000
mean      2004.605002
std         87.266189
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [111]:
autos = autos[autos["registration_year"].between(1996,2016)]
autos['registration_year'].describe()

count    41348.000000
mean      2004.282142
std          5.089302
min       1996.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [112]:
autos["registration_year"].value_counts(normalize=True)

2000    0.076279
2005    0.070886
1999    0.070040
2003    0.065203
2004    0.065082
2006    0.064380
2001    0.063679
2002    0.060003
1998    0.057028
2007    0.054537
2008    0.052868
2009    0.049990
1997    0.047088
2011    0.037414
2010    0.037342
1996    0.033158
2016    0.029022
2012    0.028828
2013    0.017147
2014    0.013374
2015    0.006651
Name: registration_year, dtype: float64

<a id='section4'></a>
# 4 Exploration of the Car Brands

In [113]:
autos['brand'].value_counts().shape

(40,)

In [114]:
brands = autos['brand'].value_counts(normalize=True)
brands

volkswagen        0.203323
opel              0.111009
bmw               0.108711
mercedes_benz     0.091588
audi              0.084792
ford              0.073716
renault           0.051345
peugeot           0.032432
fiat              0.026869
seat              0.019904
skoda             0.018235
smart             0.015938
mazda             0.015744
nissan            0.015575
citroen           0.015164
toyota            0.013253
hyundai           0.011222
mini              0.009674
volvo             0.008682
mitsubishi        0.008561
kia               0.007908
honda             0.007183
alfa_romeo        0.006796
sonstige_autos    0.006361
suzuki            0.006022
chevrolet         0.005079
chrysler          0.003507
dacia             0.002975
daihatsu          0.002660
porsche           0.002588
subaru            0.002056
jeep              0.001886
land_rover        0.001693
daewoo            0.001693
saab              0.001451
rover             0.001451
jaguar            0.001403
l

In [115]:
brands[brands>0.01].shape

(17,)

There are 40 brands in the dataset, with 17 out of them that have over a 1% of the total values, but German cars -specially Volkswagen are much more popular. So we'll focus on those ones to calculate their average price.

In [116]:
selected_brands = brands[brands>0.01]

In [117]:
avg_price_per_brand = {}

for brand in selected_brands.index:
    avg_price_per_brand[brand] = int(autos[autos['brand']==brand]['price'].mean())

In [118]:
avg_price_per_brand

{'volkswagen': 5654,
 'opel': 3097,
 'bmw': 7931,
 'mercedes_benz': 7850,
 'audi': 8613,
 'ford': 3487,
 'renault': 2476,
 'peugeot': 3165,
 'fiat': 2776,
 'seat': 4492,
 'skoda': 6408,
 'smart': 3587,
 'mazda': 4119,
 'nissan': 4976,
 'citroen': 3653,
 'toyota': 5222,
 'hyundai': 5408}

Audi, BMW and Mercdedes Benz are the most expensive cars. Then they are followed by Skoda and Japanese cars (Nissan, Toyota, Hyundai).<br>
Other non-German cars are cheaper, and Volswagen falls in between, which should have made its popularity. <br>

For these brands, let's try to understand the average mileage for those cars and if there's any visible link with mean price.

In [119]:
avg_mileage_per_brand = {}

for brand in selected_brands.index:
    avg_mileage_per_brand[brand] = int(autos[autos['brand']==brand]['odometer_km'].mean())

avg_mileage_per_brand

{'volkswagen': 128431,
 'opel': 129928,
 'bmw': 133908,
 'mercedes_benz': 132253,
 'audi': 131213,
 'ford': 125684,
 'renault': 128372,
 'peugeot': 126972,
 'fiat': 118631,
 'seat': 121609,
 'skoda': 111452,
 'smart': 99248,
 'mazda': 124861,
 'nissan': 116521,
 'citroen': 120223,
 'toyota': 115036,
 'hyundai': 106120}

In [120]:
top_brands = pd.DataFrame(pd.Series(avg_price_per_brand).sort_values(ascending=False), columns=['mean_price'])
top_brands['mean_mileage'] = pd.Series(avg_mileage_per_brand)
top_brands

mean_price  mean_mileage
audi                 8613        131213
bmw                  7931        133908
mercedes_benz        7850        132253
skoda                6408        111452
volkswagen           5654        128431
hyundai              5408        106120
toyota               5222        115036
nissan               4976        116521
seat                 4492        121609
mazda                4119        124861
citroen              3653        120223
smart                3587         99248
ford                 3487        125684
peugeot              3165        126972
opel                 3097        129928
fiat                 2776        118631
renault              2476        128372

There is no clear correlation bewteen the price and the mileage. We can even notice that the top 3 expensive cars have more mileage. But anyway almost all cars from the list have similar mileage. So we can't deduce the price from it. <br> <br>
However, it would be worth checking if cars with damage are cheaper than their non-damaged counterparts. Let's look at that now.

<a id='section5'></a>
# 5 Impact of the Damaged Status on the Price

In [121]:
autos.info()
autos['unrepaired_damage'].value_counts()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41348 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        41348 non-null  object 
 1   name                41348 non-null  object 
 2   seller              41348 non-null  object 
 3   offer_type          41348 non-null  object 
 4   price               41348 non-null  float64
 5   ab_test             41348 non-null  object 
 6   vehicle_type        39042 non-null  object 
 7   registration_year   41348 non-null  int64  
 8   gearbox             39615 non-null  object 
 9   power_ps            41348 non-null  int64  
 10  model               39658 non-null  object 
 11  odometer_km         41348 non-null  float64
 12  registration_month  41348 non-null  int64  
 13  fuel_type           38619 non-null  object 
 14  brand               41348 non-null  object 
 15  unrepaired_damage   34411 non-null  object 
 16  ad_c

nein    30493
ja       3918
Name: unrepaired_damage, dtype: int64

Out of 41348 cars, around 7000 cars have an unknown damage status. Let's remove them.

In [122]:
autos = autos[autos['unrepaired_damage'].notnull()]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34411 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        34411 non-null  object 
 1   name                34411 non-null  object 
 2   seller              34411 non-null  object 
 3   offer_type          34411 non-null  object 
 4   price               34411 non-null  float64
 5   ab_test             34411 non-null  object 
 6   vehicle_type        33432 non-null  object 
 7   registration_year   34411 non-null  int64  
 8   gearbox             33694 non-null  object 
 9   power_ps            34411 non-null  int64  
 10  model               33415 non-null  object 
 11  odometer_km         34411 non-null  float64
 12  registration_month  34411 non-null  int64  
 13  fuel_type           33010 non-null  object 
 14  brand               34411 non-null  object 
 15  unrepaired_damage   34411 non-null  object 
 16  ad_c

In [123]:
no_unrepaired = autos['unrepaired_damage'].value_counts()

In [124]:
avg_price_unrepaired ={}

for status in no_unrepaired.index:
    avg_price_unrepaired[status] = int(autos[autos['unrepaired_damage']==status]['price'].mean())

avg_price_unrepaired

{'nein': 6430, 'ja': 2323}

The cars that are damaged are in average \\$6000 cheaper than thei non-damaged counterparts. <br> 
But we should be very careful with these numbers, as there are many more undamaged cars than damaged cars. Also, the average of \\$6430 seems quite high when we look at the earlier results we found on the top brands. So the real gap between both categoris may be lower in reality.

<a id='conclusion'></a>
# 6 Conclusion

In this project, we analysed dataset on car listings from German eBay website. We did some cleaning work before looking at few factors that could impact the price.
It seems that the car brands mainly influence the price, with Volswagen being the most popular car.